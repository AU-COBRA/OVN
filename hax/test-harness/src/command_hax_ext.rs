use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandHaxExt {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

/// Computes a list of arguments that setup the number of parallel
/// jobs for dune accordingly to environment variable `DUNEJOBS`.
fn dune_jobs_args() -> Vec<String> {
    if let Ok(jobs) = std::env::var("DUNEJOBS") {
        vec!["-j".into(), jobs]
    } else {
        vec![]
    }
}

impl CommandHaxExt for Command {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_hax: PathBuf,
        }
        const CARGO_HAX: &str = "cargo-hax";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built. Note this doesn't
                // include `hax-engine-names-extract`: its build
                // script requires the driver and CLI of `hax` to be
                // available.
                assert!(Command::new("cargo")
                            .args(&["build", "--bins"])
                            .current_dir(&root)
                            .status()
                            .unwrap()
                            .success());
                let cargo_hax = cargo_bin(CARGO_HAX);
                let driver = cargo_bin("driver-hax-frontend-exporter");
                // Now the driver & CLI are installed, call `cargo
                // build` injecting their paths
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bin", "hax-engine-names-extract"])
                        .env("HAX_CARGO_COMMAND_PATH", &cargo_hax)
                        .env("HAX_RUSTC_DRIVER_BINARY", &driver)
                        .current_dir(&root)
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .args(dune_jobs_args())
                        .env("HAX_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("hax-export-json-schemas"))
                        .env("HAX_ENGINE_NAMES_EXTRACT_BINARY", cargo_bin("hax-engine-names-extract"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver,
                    cargo_hax,
                    engine: engine_dir.join("_build/install/default/bin/hax-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_hax.clone());
                cmd.env("HAX_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("HAX_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_HAX),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
