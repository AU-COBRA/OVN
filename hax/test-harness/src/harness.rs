mod command_hax_ext;
use command_hax_ext::*;
use serde_json::{Map, Value};
use std::process::{Command, Stdio};

#[derive(Clone, Debug, serde::Serialize)]
pub enum TestKind {
    Translate { backend: String },
}

impl TestKind {
    fn as_name(&self) -> String {
        (match self {
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        })
        .join("-")
    }
}

#[allow(dead_code)]
fn bool_true() -> bool {
    true
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSnapshot {
    #[serde(default = "bool_true")]
    pub stderr: bool,
    #[serde(default = "bool_true")]
    pub stdout: bool,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// is the test optional? (useful for slow tests for instance)
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
    pub snapshot: TestSnapshot,
    pub include_flag: Option<String>,
    pub backend_options: Option<Vec<String>>,
}

impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        fn as_opt_bool(v: &Value, def: bool) -> Option<bool> {
            if v.is_null() {
                return Some(def);
            }
            v.as_bool()
        }
        fn as_bool(o: &Value, k: &str, def: bool) -> bool {
            let v = &o[k];
            as_opt_bool(v, def)
                .expect(format!("[{}] was expected to be a boolean, got {}", k, v).as_str())
        }
        let snapshot = &o["snapshot"];
        TestSpec {
            optional: as_bool(&o, "optional", false),
            broken: as_bool(&o, "broken", false),
            positive: as_bool(&o, "positive", true),
            issue_id: o["positive"].as_u64(),
            include_flag: o["include-flag"].as_str().map(|s| s.into()),
            backend_options: serde_json::from_value(o["backend-options"].clone()).unwrap(),
            snapshot: as_opt_bool(snapshot, true)
                .map(|b| TestSnapshot {
                    stderr: b,
                    stdout: b,
                })
                .or_else(|| match snapshot.as_str() {
                    Some(v @ ("stdout" | "stderr" | "both" | "none")) => Some(TestSnapshot {
                        stdout: matches!(v, "stdout" | "both"),
                        stderr: matches!(v, "stderr" | "both"),
                    }),
                    Some(v) => panic!(
                        "[snapshot] is \"{}\" but was expected to be \"stderr\", \"stdout\" or \"both\"", v
                    ),
                    None => None,
                })
                .unwrap_or_else(|| TestSnapshot {
                    stderr: as_bool(&snapshot, "stderr", true),
                    stdout: as_bool(&snapshot, "stdout", true),
                }),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug, serde::Serialize)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl Test {
    fn as_args(&self) -> Vec<String> {
        match &self.kind {
            TestKind::Translate { backend } => {
                let mut args = vec![];
                args.push("into".to_string());
                if let Some(i) = self.spec.include_flag.as_ref() {
                    args.push("-i".to_string());
                    args.push(i.to_string());
                }
                args.push("--dry-run".to_string());
                args.push(backend.clone());
                if let Some(backend_options) = &self.spec.backend_options {
                    args.extend_from_slice(backend_options.clone().as_slice());
                }
                args
            }
        }
    }
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::hax(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.as_args());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            use lazy_static::lazy_static;
            use regex::Regex;
            lazy_static! {
                // Regex [TIME] matches compile times
                static ref TIME: Regex = Regex::new(r"\bin \d+(\.\d+)?s\b").unwrap();
                static ref LOCK: Regex = Regex::new(r"Blocking waiting for \w+ lock on (the registry index|build directory|package cache)").unwrap();
            }
            TIME.replace_all(
                LOCK.replace_all(
                    &s.replace(r"\", "/").replace(&workspace, "WORKSPACE_ROOT"),
                    "",
                )
                .as_ref(),
                "in XXs",
            )
            .trim()
            .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = String::from_utf8_lossy(&out.stdout).to_string();

        // 3. make sure the test is successful
        let mut snapshot: Map<String, Value> = Map::new();
        if self.spec.snapshot.stderr {
            snapshot.insert("stderr".to_string(), Value::String(serr.clone()));
        }
        if self.spec.snapshot.stdout {
            snapshot.insert(
                "stdout".to_string(),
                serde_json::from_str(&sout)
                    .map(|out: hax_types::engine_api::Output| {
                        use serde_json::json;
                        json!({
                            "diagnostics": Value::Array(out.diagnostics.into_iter().map(|diag| json!({
                                "spans": Value::Array(diag.span.clone().into_iter().map(|span| Value::String(format!("{:?}", span))).collect()),
                                "message": Value::String(format!("{}", diag)),
                            })).collect()),
                            "files": Value::Object(out.files.into_iter().map(|file| (file.path, Value::String(file.contents))).collect())
                        })
                    })
                    .unwrap_or_else(|_| Value::String(cleanup(sout.clone()))),
            );
        }

        if !snapshot.is_empty() {
            let exit = out.status.code().unwrap_or(std::i32::MAX);
            snapshot.insert("exit".to_string(), exit.into());
            let snapshot = Value::Object(snapshot);
            let name = format!("{} {}", self.info.name, self.kind.as_name());

            let mut info = self.clone();
            info.info.manifest = info.info.manifest.strip_prefix(workspace).unwrap().into();
            insta::with_settings!({
                info => &info,
            }, { insta::assert_toml_snapshot!(name, snapshot) })
        }

        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (true, (true, false) | (false, true)) => Ok(()),
            (false, (false, false) | (true, true)) => Ok(()),
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.hax-tests], this function returns a list of tests
fn parse_hax_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    if metadata.is_null() {
        return vec![];
    }

    metadata
        .as_object()
        .expect(
            format!(
                "Expected value at key [hax-tests] to be a dictionary for package {:#?}",
                info
            )
            .as_str(),
        )
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(
                    format!(
                        "Expected value at key [{}] be a dictionary for package {:#?}",
                        a, info
                    )
                    .as_str(),
                )
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                _ => panic!(
                    "unexpected metadata [hax-tests.{}.{}] for package {:#?}",
                    a, b, info
                ),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    let mut args = libtest_mimic::Arguments::from_args();
    args.test_threads = Some(1);

    libtest_mimic::run(
        &args,
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_hax_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["hax-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
