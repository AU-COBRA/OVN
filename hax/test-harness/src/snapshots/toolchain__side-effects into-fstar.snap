---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: fstar
  info:
    name: side-effects
    manifest: side-effects/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: true
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0
stderr = 'Finished `dev` profile [unoptimized + debuginfo] target(s) in XXs'

[stdout]
diagnostics = []

[stdout.files]
"Side_effects.fst" = '''
module Side_effects
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

/// Helper function
let add3 (x y z: u32) : u32 =
  Core.Num.impl__u32__wrapping_add (Core.Num.impl__u32__wrapping_add x y <: u32) z

/// Question mark without error coercion
let direct_result_question_mark (y: Core.Result.t_Result Prims.unit u32)
    : Core.Result.t_Result i8 u32 =
  match y with
  | Core.Result.Result_Ok _ -> Core.Result.Result_Ok 0y <: Core.Result.t_Result i8 u32
  | Core.Result.Result_Err err -> Core.Result.Result_Err err <: Core.Result.t_Result i8 u32

/// Question mark with an error coercion
let direct_result_question_mark_coercion (y: Core.Result.t_Result i8 u16)
    : Core.Result.t_Result i8 u32 =
  match y with
  | Core.Result.Result_Ok hoist1 -> Core.Result.Result_Ok hoist1 <: Core.Result.t_Result i8 u32
  | Core.Result.Result_Err err ->
    Core.Result.Result_Err (Core.Convert.f_from #FStar.Tactics.Typeclasses.solve err)
    <:
    Core.Result.t_Result i8 u32

/// Exercise early returns with control flow and loops
let early_returns (x: u32) : u32 =
  Rust_primitives.Hax.Control_flow_monad.Mexception.run (let! _:Prims.unit =
        if x >. 3ul
        then
          let! hoist2:Rust_primitives.Hax.t_Never =
            Core.Ops.Control_flow.ControlFlow_Break 0ul
            <:
            Core.Ops.Control_flow.t_ControlFlow u32 Rust_primitives.Hax.t_Never
          in
          Core.Ops.Control_flow.ControlFlow_Continue (Rust_primitives.Hax.never_to_any hoist2)
          <:
          Core.Ops.Control_flow.t_ControlFlow u32 Prims.unit
        else
          Core.Ops.Control_flow.ControlFlow_Continue ()
          <:
          Core.Ops.Control_flow.t_ControlFlow u32 Prims.unit
      in
      let! x, hoist5:(u32 & u32) =
        if x >. 30ul
        then
          match true with
          | true ->
            let! hoist4:Rust_primitives.Hax.t_Never =
              Core.Ops.Control_flow.ControlFlow_Break 34ul
              <:
              Core.Ops.Control_flow.t_ControlFlow u32 Rust_primitives.Hax.t_Never
            in
            Core.Ops.Control_flow.ControlFlow_Continue
            (x, Rust_primitives.Hax.never_to_any hoist4 <: (u32 & u32))
            <:
            Core.Ops.Control_flow.t_ControlFlow u32 (u32 & u32)
          | _ ->
            Core.Ops.Control_flow.ControlFlow_Continue (x, 3ul <: (u32 & u32))
            <:
            Core.Ops.Control_flow.t_ControlFlow u32 (u32 & u32)
        else
          Core.Ops.Control_flow.ControlFlow_Continue
          (let x:u32 = x +! 9ul in
            x, x +! 1ul <: (u32 & u32))
          <:
          Core.Ops.Control_flow.t_ControlFlow u32 (u32 & u32)
      in
      let! hoist8:Rust_primitives.Hax.t_Never =
        Core.Ops.Control_flow.ControlFlow_Break
        (Core.Num.impl__u32__wrapping_add (Core.Num.impl__u32__wrapping_add 123ul hoist5 <: u32) x)
        <:
        Core.Ops.Control_flow.t_ControlFlow u32 Rust_primitives.Hax.t_Never
      in
      Core.Ops.Control_flow.ControlFlow_Continue (Rust_primitives.Hax.never_to_any hoist8)
      <:
      Core.Ops.Control_flow.t_ControlFlow u32 u32)

/// Exercise local mutation with control flow and loops
let local_mutation (x: u32) : u32 =
  let y:u32 = 0ul in
  let x:u32 = Core.Num.impl__u32__wrapping_add x 1ul in
  if x >. 3ul
  then
    let x:u32 = Core.Num.impl__u32__wrapping_sub x 3ul in
    let y:u32 = x /! 2ul in
    let y:u32 = Core.Num.impl__u32__wrapping_add y 2ul in
    let y:u32 =
      Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Ops.Range.t_Range
              u32)
            #FStar.Tactics.Typeclasses.solve
            ({ Core.Ops.Range.f_start = 0ul; Core.Ops.Range.f_end = 10ul }
              <:
              Core.Ops.Range.t_Range u32)
          <:
          Core.Ops.Range.t_Range u32)
        y
        (fun y i ->
            let y:u32 = y in
            let i:u32 = i in
            Core.Num.impl__u32__wrapping_add x i <: u32)
    in
    Core.Num.impl__u32__wrapping_add x y
  else
    let (x, y), hoist15:((u32 & u32) & u32) =
      match x with
      | 12ul ->
        let y:u32 = Core.Num.impl__u32__wrapping_add x y in
        (x, y <: (u32 & u32)), 3ul <: ((u32 & u32) & u32)
      | 13ul ->
        let x:u32 = Core.Num.impl__u32__wrapping_add x 1ul in
        (x, y <: (u32 & u32)), add3 x (Core.Num.impl__u32__wrapping_add 123ul x <: u32) x
        <:
        ((u32 & u32) & u32)
      | _ -> (x, y <: (u32 & u32)), 0ul <: ((u32 & u32) & u32)
    in
    let x:u32 = hoist15 in
    Core.Num.impl__u32__wrapping_add x y

/// Test question mark on `Option`s with some control flow
let options (x y: Core.Option.t_Option u8) (z: Core.Option.t_Option u64) : Core.Option.t_Option u8 =
  Rust_primitives.Hax.Control_flow_monad.Mexception.run (match x with
      | Core.Option.Option_Some hoist19 ->
        let! hoist26:Core.Option.t_Option u8 =
          if hoist19 >. 10uy
          then
            match x with
            | Core.Option.Option_Some hoist21 ->
              Core.Ops.Control_flow.ControlFlow_Continue
              (Core.Option.Option_Some (Core.Num.impl__u8__wrapping_add hoist21 3uy)
                <:
                Core.Option.t_Option u8)
              <:
              Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8)
                (Core.Option.t_Option u8)
            | Core.Option.Option_None  ->
              Core.Ops.Control_flow.ControlFlow_Break
              (Core.Option.Option_None <: Core.Option.t_Option u8)
              <:
              Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8)
                (Core.Option.t_Option u8)
          else
            match x with
            | Core.Option.Option_Some hoist24 ->
              (match y with
                | Core.Option.Option_Some hoist23 ->
                  Core.Ops.Control_flow.ControlFlow_Continue
                  (Core.Option.Option_Some (Core.Num.impl__u8__wrapping_add hoist24 hoist23)
                    <:
                    Core.Option.t_Option u8)
                  <:
                  Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8)
                    (Core.Option.t_Option u8)
                | Core.Option.Option_None  ->
                  Core.Ops.Control_flow.ControlFlow_Break
                  (Core.Option.Option_None <: Core.Option.t_Option u8)
                  <:
                  Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8)
                    (Core.Option.t_Option u8))
            | Core.Option.Option_None  ->
              Core.Ops.Control_flow.ControlFlow_Break
              (Core.Option.Option_None <: Core.Option.t_Option u8)
              <:
              Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8)
                (Core.Option.t_Option u8)
        in
        (match hoist26 with
          | Core.Option.Option_Some hoist27 ->
            let! v:u8 =
              match hoist27 with
              | 3uy ->
                (match Core.Option.Option_None <: Core.Option.t_Option u8 with
                  | Core.Option.Option_Some some ->
                    Core.Ops.Control_flow.ControlFlow_Continue some
                    <:
                    Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) u8
                  | Core.Option.Option_None  ->
                    Core.Ops.Control_flow.ControlFlow_Break
                    (Core.Option.Option_None <: Core.Option.t_Option u8)
                    <:
                    Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) u8)
              | 4uy ->
                (match z with
                  | Core.Option.Option_Some hoist16 ->
                    Core.Ops.Control_flow.ControlFlow_Continue
                    (4uy +! (if hoist16 >. 4uL <: bool then 0uy else 3uy))
                    <:
                    Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) u8
                  | Core.Option.Option_None  ->
                    Core.Ops.Control_flow.ControlFlow_Break
                    (Core.Option.Option_None <: Core.Option.t_Option u8)
                    <:
                    Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) u8)
              | _ ->
                Core.Ops.Control_flow.ControlFlow_Continue 12uy
                <:
                Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) u8
            in
            Core.Ops.Control_flow.ControlFlow_Continue
            (match x with
              | Core.Option.Option_Some hoist28 ->
                (match y with
                  | Core.Option.Option_Some hoist29 ->
                    Core.Option.Option_Some
                    (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v hoist28
                          <:
                          u8)
                        hoist29)
                    <:
                    Core.Option.t_Option u8
                  | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
              | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
            <:
            Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) (Core.Option.t_Option u8)
          | Core.Option.Option_None  ->
            Core.Ops.Control_flow.ControlFlow_Continue
            (Core.Option.Option_None <: Core.Option.t_Option u8)
            <:
            Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) (Core.Option.t_Option u8))
      | Core.Option.Option_None  ->
        Core.Ops.Control_flow.ControlFlow_Continue
        (Core.Option.Option_None <: Core.Option.t_Option u8)
        <:
        Core.Ops.Control_flow.t_ControlFlow (Core.Option.t_Option u8) (Core.Option.t_Option u8))

/// Test question mark on `Result`s with local mutation
let question_mark (x: u32) : Core.Result.t_Result u32 u32 =
  Rust_primitives.Hax.Control_flow_monad.Mexception.run (let! x:u32 =
        if x >. 40ul
        then
          let y:u32 = 0ul in
          let x:u32 = Core.Num.impl__u32__wrapping_add x 3ul in
          let y:u32 = Core.Num.impl__u32__wrapping_add x y in
          let x:u32 = Core.Num.impl__u32__wrapping_add x y in
          if x >. 90ul
          then
            match Core.Result.Result_Err 12uy <: Core.Result.t_Result Prims.unit u8 with
            | Core.Result.Result_Ok ok ->
              Core.Ops.Control_flow.ControlFlow_Continue x
              <:
              Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result u32 u32) u32
            | Core.Result.Result_Err err ->
              let! _:Prims.unit =
                Core.Ops.Control_flow.ControlFlow_Break
                (Core.Result.Result_Err (Core.Convert.f_from #FStar.Tactics.Typeclasses.solve err)
                  <:
                  Core.Result.t_Result u32 u32)
                <:
                Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result u32 u32) Prims.unit
              in
              Core.Ops.Control_flow.ControlFlow_Continue x
              <:
              Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result u32 u32) u32
          else
            Core.Ops.Control_flow.ControlFlow_Continue x
            <:
            Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result u32 u32) u32
        else
          Core.Ops.Control_flow.ControlFlow_Continue x
          <:
          Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result u32 u32) u32
      in
      Core.Ops.Control_flow.ControlFlow_Continue
      (Core.Result.Result_Ok (Core.Num.impl__u32__wrapping_add 3ul x)
        <:
        Core.Result.t_Result u32 u32)
      <:
      Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result u32 u32)
        (Core.Result.t_Result u32 u32))

type t_A = | A : t_A

type t_B = | B : t_B

type t_Bar = {
  f_a:bool;
  f_b:(t_Array (bool & bool) (sz 6) & bool)
}

/// Combine `?` and early return
let monad_lifting (x: u8) : Core.Result.t_Result t_A t_B =
  Rust_primitives.Hax.Control_flow_monad.Mexception.run (if x >. 123uy <: bool
      then
        match Core.Result.Result_Err (B <: t_B) <: Core.Result.t_Result t_A t_B with
        | Core.Result.Result_Ok hoist33 ->
          let! hoist35:Rust_primitives.Hax.t_Never =
            Core.Ops.Control_flow.ControlFlow_Break
            (Core.Result.Result_Ok hoist33 <: Core.Result.t_Result t_A t_B)
            <:
            Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result t_A t_B)
              Rust_primitives.Hax.t_Never
          in
          Core.Ops.Control_flow.ControlFlow_Continue (Rust_primitives.Hax.never_to_any hoist35)
          <:
          Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result t_A t_B)
            (Core.Result.t_Result t_A t_B)
        | Core.Result.Result_Err err ->
          Core.Ops.Control_flow.ControlFlow_Continue
          (Core.Result.Result_Err err <: Core.Result.t_Result t_A t_B)
          <:
          Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result t_A t_B)
            (Core.Result.t_Result t_A t_B)
      else
        Core.Ops.Control_flow.ControlFlow_Continue
        (Core.Result.Result_Ok (A <: t_A) <: Core.Result.t_Result t_A t_B)
        <:
        Core.Ops.Control_flow.t_ControlFlow (Core.Result.t_Result t_A t_B)
          (Core.Result.t_Result t_A t_B))

type t_Foo = {
  f_x:bool;
  f_y:(bool & Alloc.Vec.t_Vec t_Bar Alloc.Alloc.t_Global);
  f_z:t_Array t_Bar (sz 6);
  f_bar:t_Bar
}

/// Test assignation on non-trivial places
let assign_non_trivial_lhs (foo: t_Foo) : t_Foo =
  let foo:t_Foo = { foo with f_x = true } <: t_Foo in
  let foo:t_Foo = { foo with f_bar = { foo.f_bar with f_a = true } <: t_Bar } <: t_Foo in
  let foo:t_Foo =
    {
      foo with
      f_bar
      =
      {
        foo.f_bar with
        f_b
        =
        {
          foo.f_bar.f_b with
          _1
          =
          Rust_primitives.Hax.Monomorphized_update_at.update_at_usize foo.f_bar.f_b._1
            (sz 3)
            ({ (foo.f_bar.f_b._1.[ sz 3 ] <: (bool & bool)) with _2 = true } <: (bool & bool))
        }
        <:
        (t_Array (bool & bool) (sz 6) & bool)
      }
      <:
      t_Bar
    }
    <:
    t_Foo
  in
  let foo:t_Foo =
    {
      foo with
      f_z
      =
      Rust_primitives.Hax.Monomorphized_update_at.update_at_usize foo.f_z
        (sz 3)
        ({ (foo.f_z.[ sz 3 ] <: t_Bar) with f_a = true } <: t_Bar)
    }
    <:
    t_Foo
  in
  let foo:t_Foo =
    {
      foo with
      f_y
      =
      {
        foo.f_y with
        _2
        =
        Rust_primitives.Hax.Monomorphized_update_at.update_at_usize foo.f_y._2
          (sz 3)
          ({
              (foo.f_y._2.[ sz 3 ] <: t_Bar) with
              f_b
              =
              {
                (foo.f_y._2.[ sz 3 ] <: t_Bar).f_b with
                _1
                =
                Rust_primitives.Hax.Monomorphized_update_at.update_at_usize (foo.f_y._2.[ sz 3 ]
                    <:
                    t_Bar)
                    .f_b
                    ._1
                  (sz 5)
                  ({
                      ((foo.f_y._2.[ sz 3 ] <: t_Bar).f_b._1.[ sz 5 ] <: (bool & bool)) with
                      _1 = true
                    }
                    <:
                    (bool & bool))
                <:
                t_Array (bool & bool) (sz 6)
              }
              <:
              (t_Array (bool & bool) (sz 6) & bool)
            }
            <:
            t_Bar)
      }
      <:
      (bool & Alloc.Vec.t_Vec t_Bar Alloc.Alloc.t_Global)
    }
    <:
    t_Foo
  in
  foo
'''
